### PR TITLE
Using vector<unsigned char> instead of vector<bool>

### PR DESCRIPTION
**Pull request by ChengChen002**
Since the std::vector\<bool\> type does not have a data() member function,
Using vector<unsigned char> instead of vector<bool>.

### CODE REVIEW

**Comment by ChengChen002** [comment]
I haven't come up with a cleaner way to fix this yet, please comment if you have any other suggestions! Thanks.

**Comment by bader** [comment]
> I haven't come up with a cleaner way to fix this yet, please comment if you have any other suggestions! Thanks.

Please, push "Ready for review" button if you want this PR to be reviewed.

**Review by keryell — COMMENTED** [review]
This looks correct but there is a lot of code duplication.
I was thinking about introducing a generic container type that would select either a `std::vector` or a `std::array` according to the value type, but thinking again, why not just using always a `std::array`?

**Comment by ChengChen002** [comment]
> This looks correct but there is a lot of code duplication. I was thinking about introducing a generic container type that would select either a `std::vector` or a `std::array` according to the value type, but thinking again, why not just using always a `std::array`?

This is to avoid using too much stack memory, you can see more information in #823.

**Review by keryell — APPROVED** [review]
Thanks! I love shorter code for maintenance. :-)

File: tests/usm/usm_api.h [hunk 1]
```diff
@@ -511,30 +511,34 @@ struct gen_buf_size_selector<0> {
  */
 template <typename T, size_t buf_size>
 class event_generator {
+  // Change vector<bool> to vector <unsigned char>, since vector<bool> is
+  // stored bit-wise, and does not have data() member function.
+  using ContainType = typename std::conditional<std::is_same_v<bool, T>,
+                                                unsigned char, T>::type;
```
Comment (keryell) [inline tests/usm/usm_api.h:517]: C++14
```suggestion
  using ContainType = typename std::conditional_t<std::is_same_v<bool, T>,
                                                unsigned char, T>;
```

**Review by keryell — CHANGES_REQUESTED** [review]

File: tests/usm/usm_api.h [hunk 2]
```diff
@@ -511,30 +511,34 @@ struct gen_buf_size_selector<0> {
  */
 template <typename T, size_t buf_size>
 class event_generator {
+  // Change vector<bool> to vector <unsigned char>, since vector<bool> is
+  // stored bit-wise, and does not have data() member function.
+  using ContainType = typename std::conditional_t<std::is_same_v<bool, T>,
+                                                  unsigned char, T>;
```
Comment (bader) [inline tests/usm/usm_api.h:517]:
```suggestion
  using ContainType =
      typename std::conditional_t<std::is_same_v<bool, T>, unsigned char, T>;
```

**Review by keryell — APPROVED** [review]
Thanks!

**Pull request merged**